autofish.lua

```diff
@@ -0,0 +1,220 @@
+-- handles automatic fishing jobs to limit the number of fish the fortress keeps on hand
+-- autofish [enable | disable] <max> [min] [--include-raw | -r]
+
+--@ enable=true
+
+local json = require("json")
+local persist = require("persist-table")
+local argparse = require("argparse")
+local dump = require("dumper")
+
+local GLOBAL_KEY = "autofish"
+
+local help = [====[
+autofish
+=============
+Manages your fish stocks by toggling fishing labours as you reach certain stock thresholds.
+
+
+Usage
+-------------
+  autofish [enable | disable | status] | <max> [min] [--include-raw | -r]
+
+Arguments
+-------------
+-r, --include-raw
+  Allow the stock tracking to also count raw fish
+enable
+  Enable this script
+disable
+  Disable this script
+status
+  Print the current status
+max
+  The maximum number of fish you want to stop fishing at.
+min
+  the minimum number of fish before restarting fishing.
+]====]
+-- set default enabled state
+enabled = enabled or false
+set_maxFish = set_maxFish or 100
+set_minFish = set_minFish or 50
+set_useRaw = set_useRaw or false
+isFishing = isFishing or true
+
+function isEnabled()
+    return enabled
+end
+
+local function persist_state()
+    persist.GlobalTable[GLOBAL_KEY] = json.encode({enabled=enabled,
+        set_maxFish=set_maxFish, set_minFish=set_minFish, set_useRaw=set_useRaw,
+        isFishing=isFishing
+    })
+end
+
+local function load_state()
+    -- load persistent data
+    local persisted_data = json.decode(persist.GlobalTable[GLOBAL_KEY] or "")
+    enabled = (persisted_data or {enabled=false})["enabled"]
+    set_maxFish = (persisted_data or {set_maxFish=100})["set_maxFish"]
+    set_minFish = (persisted_data or {set_minFish=50})["set_minFish"]
+    set_useRaw = (persisted_data or {set_useRaw=false})["set_useRaw"]
+    isFishing = (persisted_data or {isFishing=true})["isFishing"]
+end
+
+
+-- toggle the fishing labour on all dwarves/work detail if enabled
+function toggle_fishing_labour(state)
+    -- pass true to state to turn on, otherwise disable
+    -- find all work details that have fishing enabled:
+    local work_details = df.global.plotinfo.hauling.work_details
+    for k,v in pairs(work_details) do
+        if v.allowed_labors.FISH then
+            -- set limited to true just in case a custom work detail is being
+            -- changed, to prevent *all* dwarves from fishing.
+            v.work_detail_flags.limited = true
+            v.work_detail_flags.enabled = state
+        end
+    end
+    isFishing = state -- save current state
```

Comment:
if we have changed state, it might be nice to report that with a `print('autofish: threshold crossed; stopping fishing')` so the player can see the record on the console if they're confused why people suddenly started/stopped fishing